src/promql.grammar

```diff
@@ -1,4 +1,4 @@
-@top { Expr }
+@top PromQL { Expr }
```

Comment:
An alternative would be to omit the `PromQL` (since we know it's always PromQL anyway), and make `Expr` the top-level thing:

```
@top Expr {
  AggregateExpr |
  BinaryExpr |
  FunctionCall |
  MatrixSelector |
  NumberLiteral |
  OffsetExpr |
  ParenExpr |
  StringLiteral |
  SubqueryExpr |
  UnaryExpr |
  VectorSelector
}
```

What do you think?

Reply:
Depending if you can call the top when you are in the [BinOp](https://github.com/promlabs/lezer-promql/blob/master/src/promql.grammar#L52) for example

Reply:
and if it works, yeah we can do that. It will simplify a bit the grammar. Just a bit shame that the generated grammar doesn't create the constant `PromQL` with the corresponding number value.

So in the code, we have to test using the name when we want to check if the given node is `PromQL`

Reply:
Ah ok, if it causes problems like that, I'm happy to keep `PromQL` the top like you have here.

Reply:
So apparently, if I'm removing the rule `Expr` to have only the Top named like that, it doesn't work:

```
Reference to undefined rule 'Expr' (src/promql.grammar 168:23)
```

But if I have that:

```
@top Expr {
  AggregateExpr |
  BinaryExpr |
  FunctionCall |
  MatrixSelector |
  NumberLiteral |
  OffsetExpr |
  ParenExpr |
  StringLiteral |
  SubqueryExpr |
  UnaryExpr |
  VectorSelector
}

Expr {
  AggregateExpr |
  BinaryExpr |
  FunctionCall |
  MatrixSelector |
  NumberLiteral |
  OffsetExpr |
  ParenExpr |
  StringLiteral |
  SubqueryExpr |
  UnaryExpr |
  VectorSelector
}
```
It's building. But it's really weird. I believe doing that will force us to do the following check, in order to make the difference between the top and the node `Expr`:

``` typescript
if (node.isTop()) {
   // allow to know if the node is the top. ( coming from lezer v0.12.0)
}
```

``` typescript
if (node.type.id === Expr) {
  // we are in the node Expr which is not the top
}
```

So perhaps we could do something like that:

```
@top PromQL{
  AggregateExpr |
  BinaryExpr |
  FunctionCall |
  MatrixSelector |
  NumberLiteral |
  OffsetExpr |
  ParenExpr |
  StringLiteral |
  SubqueryExpr |
  UnaryExpr |
  VectorSelector
}

Expr {
  AggregateExpr |
  BinaryExpr |
  FunctionCall |
  MatrixSelector |
  NumberLiteral |
  OffsetExpr |
  ParenExpr |
  StringLiteral |
  SubqueryExpr |
  UnaryExpr |
  VectorSelector
}
```

It will avoid a confusion between the top and the node + it removes a step between the top and the effective node.

What do you want to do ? Keeping the way proposed by the PR, or removing a step between the top and the effective nodes

Reply:
Meh. Ok. Then I guess we should just keep it like you have it now. Thanks for the explanation!